### ISSUE TITLE
`<For>` panics when no parent is found

### ISSUE THREAD

**Issue by lpotthast** [leptos-rs/leptos#897]
The following error (originating from `each.rs:689:45`)
```
ok-14eeb0630fba0fef.js:316 panicked at 'called `Result::unwrap()` on an `Err` value: JsValue(InvalidNodeTypeError: Failed to execute 'setStartBefore' on 'Range': the given Node has no parent.
Error: Failed to execute 'setStartBefore' on 'Range': the given Node has no parent.
    at http://127.0.0.1:4001/book-14eeb0630fba0fef.js:552:21
    at handleError (http://127.0.0.1:4001/book-14eeb0630fba0fef.js:241:18)
    at imports.wbg.__wbg_setStartBefore_8e6e0f57eb922e33 (http://127.0.0.1:4001/book-14eeb0630fba0fef.js:551:73)
    at web_sys::features::gen_Range::Range::set_start_before::h46a1bcc0e6acd02c (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[6477]:0x6164ad)
    at leptos_dom::components::each::apply_cmds::h36201b78da8b02bd (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[273]:0xc2850)
    at <leptos_dom::components::each::Each<IF,I,T,EF,N,KF,K> as leptos_dom::IntoView>::into_view::{{closure}}::ha10796ab77ff0069 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[190]:0x52a29)
    at <leptos_reactive::effect::Effect<T,F> as leptos_reactive::effect::AnyComputation>::run::heb8296b5f041a68b (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[217]:0x794e3)
    at leptos_reactive::runtime::Runtime::update::{{closure}}::hf0b6181b95189308 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[6997]:0x638eaf)
    at leptos_reactive::runtime::Runtime::with_observer::he70bc21b3d6648b9 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[4561]:0x578d68)
    at leptos_reactive::runtime::Runtime::update::h607d508bb92482a6 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[828]:0x2c488f))', /home/lukas/.cargo/registry/src/github.com-1ecc6299db9ec823/leptos_dom-0.2.5/src/components/each.rs:689:45

Stack:

Error
    at imports.wbg.__wbg_new_abda76e883ba8a5f (http://127.0.0.1:4001/book-14eeb0630fba0fef.js:319:17)
    at console_error_panic_hook::Error::new::hfdaedaff5aed9273 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[17751]:0x7a5a5d)
    at console_error_panic_hook::hook_impl::hd09abaec27218d55 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[4538]:0x57699a)
    at console_error_panic_hook::hook::ha3d03e0aeec5abef (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[19642]:0x7bf546)
    at core::ops::function::Fn::call::h57386142573e6411 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[17186]:0x79cbde)
    at std::panicking::rust_panic_with_hook::h560812122b50d13d (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[8583]:0x695479)
    at std::panicking::begin_panic_handler::{{closure}}::h444de343e52388a7 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[10566]:0x6f06fe)
    at std::sys_common::backtrace::__rust_end_short_backtrace::hfd641dea4342d358 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[21070]:0x7d1333)
    at rust_begin_unwind (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[15298]:0x779acc)
    at core::panicking::panic_fmt::hfa9ae79d6c4dc316 (http://127.0.0.1:4001/book-14eeb0630fba0fef_bg.wasm:wasm-function[18828]:0x7b4ccf)
```
was observed when using the following markup
```
<Show when=my_non_changing_signal fallback=|_cx| view! { cx,  }>
    <For
        each=my_changing_signal
        key=|my_entry| my_entry.id
        view=move |cx, my_entry| view! { cx,
            <div>"seems irrelevant"</div>
        }
    />
</Show>
```
The error no longer appeared after wrapping the `<For>` element in an explicit `<div>`. This should probably not be required in this situation.
It seems that the calls to `set_start_before` and `set_end_before`
```
    range.set_start_before(opening).unwrap();
    range.set_end_before(closing).unwrap();
```
both require that `opening` and `closing` have a parent? This is at least is not checked right now.

**Comment by gbj** [comment]
Thanks. Agreed that it should not be required. The `<For/>` component is currently being rehabbed with some associated tests and clean-up due to an off-by-one error in a particular case (#533) so I'll keep this on my list to check when that work is done.

**Comment by lpotthast** [comment]
Sounds great!

**Comment by phillipbaird** [comment]
This has tripped me up as well.
[Here's a sample project](https://github.com/phillipbaird/leptos-troubleshoot) to demonstrate the problem.

**Comment by gbj** [comment]
I'm no longer able to reproduce this against either of the provided examples. It may have been related to #1078. In any case, I'm going to close this as fixed... Let me know if you're still encountering this issue against latest git main and I can reopen it.

**Comment by Sett17** [comment]
This error still occurs for me. Although I can't quite explain when it happens

When pressing enter in the input, from the markup below, everything works fine and the tag is added.
When the `tags` signal is updated from somewhere else, one of two things happen:

1. **First time update (For was empty until now)** Everything works as expected. Tags are added
2. **Update after previous update (either by button click or input enter)** The error occurs, and then For is not usable anymore

I am not 100% sure if this is still an error in Leptos or if I am using the `For` wrong in any way.

The markup:
```rust
<div class="flex-grow px-3 py-1 text-white bg-black border rounded-lg focus:outline-none focus:shadow-outline flex flex-row h-10 mb-3">
    <For
        each=tags
        key=|item: &String| item.clone()
        view=|cx, item: String| {
            view! { cx,
                <span class="rounded-md border p-1 hover:bg-metroid cursor-pointer mx-1 text-xs flex items-center"
                    on:click=move |ev| {
                        let d = ev.target().expect("No target").dyn_into::<web_sys::HtmlSpanElement>().unwrap().inner_text();
                        // set_tags1.update(|s|{ //this also does not work, but that's a topic for another discussion
                        //     s.0.retain(|t| t != &d);
                        // });
                    }
                >
                    {item}
                </span>
            }
        }
    />
    <input
        class="text-white placeholder-gray-400 bg-black border-0 focus:outline-none flex-grow ml-1"
        type="text"
        prop:placeholder=move || {if tags().0.len() > 0 { "Tags separated by <enter>" } else { "Tags separated by <enter> (required)" }}
        on:keydown=move |ev| {
            if ev.key() == "Enter" {
                let value = event_target_value(&ev);
                if value.len() > 0 {
                    set_tags2.update(|s|{
                        s.0.push(value);
                    });
                    ev.prevent_default();
                    ev.target().unwrap().dyn_into::<HtmlInputElement>().unwrap().set_value("");
                }
            }
        }
    />
</div>
```

<details><summary>The error</summary>
<pre>
panicked at 'called `Result::unwrap()` on an `Err` value: JsValue(InvalidNodeTypeError: Failed to execute 'setStartBefore' on 'Range': the given Node has no parent.
Error: Failed to execute 'setStartBefore' on 'Range': the given Node has no parent.
    at http://localhost:3000/pkg/snac.js:655:17
    at handleError (http://localhost:3000/pkg/snac.js:295:38)
    at imports.wbg.__wbg_setStartBefore_16768752de106d95 (http://localhost:3000/pkg/snac.js:654:73)
    at web_sys::features::gen_Range::Range::set_start_before::h96f942a4e2a14dcb (http://localhost:3000/pkg/snac.wasm:wasm-function[4449]:0x58461c)
    at leptos_dom::components::each::apply_cmds::he3cc9519592ff309 (http://localhost:3000/pkg/snac.wasm:wasm-function[284]:0x10eb87)
    at <leptos_dom::components::each::Each<IF,I,T,EF,N,KF,K> as leptos_dom::IntoView>::into_view::{{closure}}::h6036959e2e067b24 (http://localhost:3000/pkg/snac.wasm:wasm-function[279]:0x103fdf)
    at <leptos_reactive::effect::Effect<T,F> as leptos_reactive::effect::AnyComputation>::run::hcb3fb0c0d1a5e9f2 (http://localhost:3000/pkg/snac.wasm:wasm-function[417]:0x206bd1)
    at leptos_reactive::runtime::Runtime::update::{{closure}}::h1a1429a704e6a2fb (http://localhost:3000/pkg/snac.wasm:wasm-function[5303]:0x5b8388)
    at leptos_reactive::runtime::Runtime::with_observer::hb8ef2bfe4199d859 (http://localhost:3000/pkg/snac.wasm:wasm-function[5563]:0x5c6586)
    at leptos_reactive::runtime::Runtime::update::h5b55679c2787f923 (http://localhost:3000/pkg/snac.wasm:wasm-function[854]:0x3943db))', /home/dikka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/leptos_dom-0.3.0/src/components/each.rs:692:45

Stack:

Error
    at imports.wbg.__wbg_new_abda76e883ba8a5f (http://localhost:3000/pkg/snac.js:370:13)
    at console_error_panic_hook::Error::new::h53975328d482b6c9 (http://localhost:3000/pkg/snac.wasm:wasm-function[12113]:0x683c8a)
    at console_error_panic_hook::hook_impl::hc406df7ae9167657 (http://localhost:3000/pkg/snac.wasm:wasm-function[2841]:0x501116)
    at console_error_panic_hook::hook::h2f56a5219f622277 (http://localhost:3000/pkg/snac.wasm:wasm-function[13112]:0x691583)
    at core::ops::function::Fn::call::h49a136941c8dd361 (http://localhost:3000/pkg/snac.wasm:wasm-function[11617]:0x67c1b0)
    at std::panicking::rust_panic_with_hook::h6ad5e2625393b2cc (http://localhost:3000/pkg/snac.wasm:wasm-function[5670]:0x5cbfc6)
    at std::panicking::begin_panic_handler::{{closure}}::h93e72541ffd61b32 (http://localhost:3000/pkg/snac.wasm:wasm-function[7115]:0x60d67b)
    at std::sys_common::backtrace::__rust_end_short_backtrace::h9d1952ad96bb3372 (http://localhost:3000/pkg/snac.wasm:wasm-function[15250]:0x6a5b6c)
    at rust_begin_unwind (http://localhost:3000/pkg/snac.wasm:wasm-function[10093]:0x66019f)
    at core::panicking::panic_fmt::h93b49f36550e1ecf (http://localhost:3000/pkg/snac.wasm:wasm-function[12844]:0x68dedd)


</pre>
</details>

**Comment by lpotthast** [comment]
Are you using the master branch or the released version of Leptos (0.3)? The `For` component of 0.3 still has many problems.

**Comment by Sett17** [comment]
That's it. I didn't check if any of the fixes mentioned here made it into a release yet.
